### DISCUSSION TITLE
Is there a way to have an indicator referring to the interface that a function implements and to see where an interface is implemented?

### DISCUSSION THREAD

**Discussion by clauderoy790** [golang/vscode-go#1735] [Help] [answered]
I decided to make the switch from GoLand to VS Code and I'm liking it a lot so far. One of the thing that I haven't found how to do yet is VS Code that is possible to do in GoLand is to have an indicator next to a function that implements an interface method that refers to the implemented interface as such: 

![image](https://user-images.githubusercontent.com/28652077/131478732-4954a95b-6cea-448e-a3bc-a1e77c9f745b.png)

Also having an indicator in the interface itself that refers to implementation of that same interface.
![image](https://user-images.githubusercontent.com/28652077/131478900-f07686b5-4579-4f04-adf2-a1e4e3cc75c4.png)

Does anyone know if there's an extension for that in VS Code? I know it's possible to go to implementations from the interface when you right click it -> Go to implementations and from the interface implementation it works as well, but having an indicator is super useful because you see it right away.

If it doesn't exist, does anyone know if it'd be difficult to make an extension for that? I've never made one, but I'd be 100% done for it!

Thanks for the help,

Claude

**Comment by hyangah** [comment, marked as answer]
If you run "Find All Implementations" **from the struct type**, vscode will show the list of interfaces the type implements. 
If you run "Find All Implementations" **from a method**, vscode will show the list of corresponding interfaces.
This is not very intuitive but it looks like this was an attempt to fit Go-specific information in the general purpose LSP concept back then. 
https://github.com/golang/go/issues/35550 

There is no visual indicator though.

**Reply by clauderoy790** [reply]
Makes sense, thanks for clarifying that out!

**Reply by guettli** [reply]
BTW, there is an issue about getting up-arrow/down-arrow on interfaces and methods implementing an interface (like Golang): https://github.com/golang/go/issues/56695 - maybe it is coming soon!

**Reply by tony-sol** [reply]
So is this not planned as part of plugin at all?

**Comment by jcastilloa** [comment]
I agree, I was looking for the same thing. These are those small features that once you get used to them, you can't live without.
I would switch to VSCode if it had more convenient interface navigation and maybe I didn't need a microscope to interpret the source tree, which is another thing that I don't find very usable, but that's another topic :))

**Comment by XiaoXuxxxx** [comment]
I ran into the same issue and ended up making a small VS Code extension that shows a CodeLens with the number of implementations above interfaces and methods. It uses `vscode.commands.executeCommand` to run the built-in `vscode.executeImplementationProvider` command repeatedly. And when you click on the CodeLens, it shows the implementation locations, just like when you use the context menu “Go to Implementations” (Ctrl+F12).

You can try it out on the Marketplace https://marketplace.visualstudio.com/items?itemName=xiaoxuxxxx.gogo-codelens

or build the .vsix locally from here https://github.com/xiaoxuxxxx/gogo-codelens?tab=readme-ov-file#option-2-build-and-install-locally

here's a screenshot
![img](https://github.com/user-attachments/assets/30994841-a509-4449-9972-840b12be76af)

**Reply by guettli** [reply]
thank you, works fine!

**Reply by BharathMC** [reply]
Thank you!

**Reply by wylswz** [reply]
Awesome man! Somebody plz add this feature to official plugin!

**Comment by iambpn** [comment]
Someone have already created this so its doable and help devs, so why can't we have it in the official go extension?

**Reply by firelizzard18** [reply]
See https://github.com/golang/go/issues/56695 and https://go-review.googlesource.com/c/vscode-go/+/619115. **Doable** is not enough. For it to be included in gopls/vscode-go, it has to be able to scale to large projects and still perform well. That isn't easy.
